src/EditorFeatures/CSharp/EventHookup/EventHookupCommandHandler_TabKeyCommand.cs

```diff
@@ -85,116 +78,179 @@ private void HandleTabWorker(ITextView textView, ITextBuffer subjectBuffer, Acti
             }
         }
 
-        // Blocking wait (if necessary) to determine whether to consume the tab and
-        // generate the event handler.
-        EventHookupSessionManager.CurrentSession.GetEventNameTask.Wait(cancellationToken);
+        var subjectBuffer = args.SubjectBuffer;
+        var caretPoint = args.TextView.GetCaretPoint(subjectBuffer);
+        if (caretPoint is null)
+            return false;
+
+        var currentSnapshot = subjectBuffer.CurrentSnapshot;
+        var document = currentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+        if (document is null)
+            return false;
+
+        // Now emit the event asynchronously.
+        var token = _asyncListener.BeginAsyncOperation(nameof(ExecuteCommand));
+
+        // Capture everything we need off of the session manager as we'll be dismissing the core session immediately
+        // before we kick off the work to emit hte event.
+        var eventNameTask = EventHookupSessionManager.CurrentSession.GetEventNameTask;
+        var applicableToSpan = EventHookupSessionManager.CurrentSession.TrackingSpan.GetSpan(currentSnapshot);
+
+        // Ensure no matter what that once tab is hit that we're back to the initial no-session state. We do not
+        // want to cancel the bg tasks kicked off as we need their values to actually emit the event.
+        EventHookupSessionManager.DismissExistingSessions(cancelBackgroundTasks: false);
```

Comment:
this dismissed the `(Press TAB to insert)` tooltip.  we'll then pop up the `Generating event...` tooltip below in the BWI.